entrypoint

```diff
@@ -194,6 +194,17 @@ else
       done
     fi
 
+    # Execute a script to modify the name of the target file before uploading to the dump target.
+    # For example, uploading to a date stamped object key path in S3, i.e. s3://bucket/2018/08/23/path
+    if [ -f /scripts.d/target.sh ] && [ -x /scripts.d/target.sh ]; then
+        TARGET=$(NOW=${now} DUMPFILE=${TMPDIR}/${TARGET} DUMPDIR=${uri[path]} DB_DUMP_DEBUG=${DB_DUMP_DEBUG} /scripts.d/target.sh)
```

Comment:
If you do this, then the upload (`s3`, `file`, `smb`) is going to use `TARGET` not just for the _remote_ filename, but also for the local. Isn't that a problem?

E.g. `TARGET` is `db_backup_20180824162435.gz`, so the file is `${TMPDIR}/db_backup_20180824162435.gz`. You plan to upload to `s3`. The backup runs `target.sh` and gets `2018/08/24/backup_162435.gz`. Rather than uploading `${TMPDIR}/db_backup_20180824162435.gz` to `s3://bucket/2018/08/24/backup_162435.gz`, it will try to upload `${TMPDIR}/2018/08/24/backup_162435.gz` to `s3://bucket/2018/08/24/backup_162435.gz`, which will fail, since `${TMPDIR}/2018/08/24/backup_162435.gz` does not exist.

Reply:
> If you do this, then the upload (s3, file, smb) is going to use TARGET not just for the remote filename, but also for the local. Isn't that a problem?

True, but there are 2 sides to this:

- We introduce a separate variable for the remote filename so we'll have a different local and remote filename OR
- The responsibility for any changes related to the modification of the target filename rests on the end-user of the tool. So in the example you provided, if they do change the target name to ```2018/08/24/backup_162435.gz```, they'll have to make sure that that folder path exists locally.

Ideally, option a) does prevent the end-user from tripping up and forgetting to handle any side-effects of their filename manipulation, so I can make some updates to that effect. What do you think?

Reply:
There are 3 states to consider:

* What the output from the backup itself should be. This _always_ should be our default `TARGET`
* What final file "uploader" should look for, i.e. the "from" of the uploader. By default, this is the same as `TARGET`.
* What filename the "uploader" should use on the remote end, i.e. the "two" of the uploader. By default, this is the same as `TARGET`.

Defaults are good, but overrides that assume some partial default is asking for trouble, especially, as you pointed out, if the post-backup does some file manipulation that assumes something about the uploader.

Thus, I think you are right, we should be able to specify options both for local and remote part of the uploader. 

Let's have _two_ scripts in `/scripts.d/`:

* `/scripts.d/source.sh` - the name of the file that is the "from". By default, this is the original `TARGET`. If your post-backup changes the file, and you don't have `source.sh`, you will fail. But you have a way to change it now.
* `/scripts.d/target.sh` - the name of the file that is the "to"

Works?

Reply:
In the meantime, did you finish the update? Once this works and passes CI, we can merge it in, and deal with your test run issue separately.

Reply:
> In the meantime, did you finish the update?

I introduced the concept of source and target and wanted to test that the original library would continue to work as is, before proceeding with the script overrides. Because I couldn't get the tests to work, I halted further development.

I'll complete the implementation and add new tests to see if travis passes.